File: app/src/main/java/org/scp/gymlog/ui/createexercise/CreateExerciseFormRecyclerViewAdapter.java
package org.scp.gymlog.ui.createexercise;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.databinding.ListElementFragmentFormBinding;

import java.util.List;

public class CreateExerciseFormRecyclerViewAdapter extends RecyclerView.Adapter<CreateExerciseFormRecyclerViewAdapter.ViewHolder> {

	private final List<FormElement> formElements;

	public CreateExerciseFormRecyclerViewAdapter(List<FormElement> formElements) {
		this.formElements = formElements;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		return new ViewHolder(
				ListElementFragmentFormBinding.inflate(
						LayoutInflater.from(parent.getContext()), parent, false
				)
		);
	}

	@Override
	public void onBindViewHolder(final ViewHolder holder, int position) {
		holder.formElement = formElements.get(position);
		holder.mTitleView.setText(holder.formElement.getTitle());
		holder.updateValue();
		holder.formElement.setUpdateListener(holder::updateValue);
	}

	@Override
	public int getItemCount() {
		return formElements.size();
	}

	public class ViewHolder extends RecyclerView.ViewHolder {
		public FormElement formElement;
		public final TextView mTitleView;
		public final TextView mContentView;
		public final ImageView mImageView;

		public ViewHolder(ListElementFragmentFormBinding binding) {
			super(binding.getRoot());
			mTitleView = binding.title;
			mContentView = binding.content;
			mImageView = binding.image;

			itemView.setOnClickListener(v ->
				formElement.onClick(v)
			);
		}

		public void updateValue() {
			mImageView.setImageDrawable(formElement.getDrawable());

			if (formElement.getValueStr() != null && !formElement.getValueStr().isEmpty())
				mContentView.setText(formElement.getValueStr());
			else if (formElement.getValue() == 0)
				mContentView.setText("-");
			else
				mContentView.setText(formElement.getValue());
		}

		@Override
		public String toString() {
			return super.toString() + " '" + mContentView.getText() + "'";
		}
	}
}

File: app/src/main/java/org/scp/gymlog/ui/createexercise/CreateExerciseFormRecyclerViewAdapter.kt
package org.scp.gymlog.ui.createexercise

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.scp.gymlog.databinding.ListElementFragmentFormBinding

class CreateExerciseFormRecyclerViewAdapter(
	private val formElements: List<FormElement>
) :	RecyclerView.Adapter<CreateExerciseFormRecyclerViewAdapter.ViewHolder>() {

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
		return ViewHolder(
			ListElementFragmentFormBinding.inflate(
				LayoutInflater.from(parent.context), parent, false
			)
		)
	}

	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
		val element = formElements[position];

		holder.formElement = element
		holder.mTitleView.setText(element.title)
		holder.updateValue()
		element.updateListener = Runnable { holder.updateValue() }
	}

	override fun getItemCount(): Int {
		return formElements.size
	}

	inner class ViewHolder(binding: ListElementFragmentFormBinding) :
		RecyclerView.ViewHolder(binding.root) {
		var formElement: FormElement? = null
		val mTitleView: TextView = binding.title
		private val mContentView: TextView = binding.content
		private val mImageView: ImageView = binding.image

		init {
			itemView.setOnClickListener { view -> formElement?.onClick(view) }
		}

		fun updateValue() {
			mImageView.setImageDrawable(formElement!!.drawable)

			if (formElement!!.valueStr.isNotEmpty())
				mContentView.text = formElement!!.valueStr
			else if (formElement!!.value == 0)
				mContentView.text = "-"
			else
				mContentView.setText(formElement!!.value)
		}

		override fun toString(): String {
			return super.toString() + " '" + mContentView.text + "'"
		}
	}
}